bootloader/src/pyi_archive.c

```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
 #endif /* ifdef _WIN32 */
 }
 
+/*
+ * Try to find the pydata section in an ELF executable.
+ * Returns the end of the section if found.
+ */
+static int
+findElfSection(ARCHIVE_STATUS * const status)
+{
+#if !defined(WIN32) && !defined(__APPLE__)
+    char head[64];
+    uint16_t endianess_bytes = 0x0201;
+    int8_t bits;
+    uint64_t e_shoff;
+    uint16_t e_shentsize;
+    uint16_t e_shnum;
+    uint16_t e_shstrndx;
+    char *section;
+    uint64_t shstr_offset;
+    uint64_t shstr_size;
+    char *shstr;
+    uint32_t section_idx;
+    uint64_t pydata_offset;
+    uint64_t pydata_size;
+
+    fseek(status->fp, 0, SEEK_SET);
+    if (fread(head, 64, 1, status->fp) != 1) {
+        return -1;
+    }
+
+    /* Check if ELF magic present */
+    if (head[0] != 0x7F || head[1] != 'E' || head[2] != 'L' || head[3] != 'F') {
+        VS("LOADER: %s is not an ELF binary\n", status->archivename);
+        return -1;
+    }
+
+    /* Ensure endianess of file matches */
+    if (head[0x05] != *(uint8_t*)&endianess_bytes) {
+        VS("LOADER: %s endianess does not match CPU\n", status->archivename);
+        return -1;
+    }
+
+    /* Ensure ELF version is supported */
+    if (head[0x06] != 1 || *(uint32_t*)(head + 0x14) != 1) {
+        VS("LOADER: %s has an unknown ELF version\n", status->archivename);
+        return -1;
+    }
+
+    /* Load ELF section header parameters */
+    bits = head[0x04]; /* 2=64, 1=32 */
+    if (bits == 2) {
+        e_shoff = *(uint64_t*)(head + 0x28);
+        e_shentsize = *(uint16_t*)(head + 0x3a);
+        e_shnum = *(uint16_t*)(head + 0x3c);
+        e_shstrndx = *(uint16_t*)(head + 0x3e);
+    } else {
+        e_shoff = *(uint32_t*)(head + 0x20);
+        e_shentsize = *(uint16_t*)(head + 0x2e);
+        e_shnum = *(uint16_t*)(head + 0x30);
+        e_shstrndx = *(uint16_t*)(head + 0x32);
```

Comment:
@htgoebel We could look into using headers. My first choice using the well known offsets was to avoid having to ifdef a bunch of headers for Linux and other Unix platforms. Without the code could compile on Windows/Apple. Regarding the ifdef comments above, what would be your preference?

At least on Linux we have Multilib, which allows us to execute 32bit binaries on a 64bit kernel and userland.

Reply:
IC, all these values are read from the file.